src/middleware/router.js

```diff
@@ -450,16 +453,69 @@ function sendHttpRequest (ctx, route, options) {
           uncompressedBodyBufs.push(data)
         })
       }
+*/
+//      const bufs = []
+ 
+      if (isNaN(counter)) {
+        counter = 0
+        size = 0
+
+        if(!bucket) {
+          bucket = getGridFSBucket()
+        }
 
-      const bufs = []
+        ctx.response.body = new Writable({
+          write: function(chunk, encoding, next) {
+            counter++
+            size += chunk.toString().length
+            console.log(`Write Response CHUNK # ${counter} upstream [ Cum size ${size} ]`)
+            next()
+          }
+        }).on('error', (err) => {
+            logger.error('Error sending Response upstream: ' + JSON.stringify(err))
+            reject(err)
+          })
+          .on('finish', () => {
+            logger.info(`Finished sending Response upstream`)
+            resolve(response)
+          })
+
+        uploadStream = bucket.openUploadStream()
+        ctx.response.bodyId = uploadStream.id
+
+        uploadStream
+          .on('error', (err) => {
+            logger.error('Storing of response in gridfs failed, error: ' + JSON.stringify(err))
+          })
+          .on('finish', (file) => {
+            logger.info(`Response body with body id: ${file._id} stored`)
+          })
+      }
 
       // See https://www.exratione.com/2014/07/nodejs-handling-uncertain-http-response-compression/
       routeRes
         .on('data', (chunk) => {
-          bufs.push(chunk)
+          if (!ctx.responseTimestamp) {
+            ctx.responseTimestamp = new Date()
+            messageStore.initiateResponse(ctx, () => {})
+          }
+          uploadStream.write(chunk)
+          ctx.response.body.write(chunk)
+          response.body.push(chunk)
```

Comment:
What is the difference between `ctx.response.body.write` and `response.body.push` here?

Reply:
I'm not sure about this, but it seems the 'ctx.response.body.write(chunk)' doesn't work. From the tests I've done, the 'response.body.push' is what works so I'm going to strip out the 'ctx.response.body.write(chunk)'. The function setKoaresponse sets the ctx.response based on the response.

Reply:
Reading up on this a bit, the `.write` doesnt seem valid and should be removed. The koa `ctx.body` / `ctx.response.body` can accept chucks being pushed to it, or it can accept an entire stream directly piped into it. In our case we need to push the chucks as we are using the chucks for other things as well. E.g `initiateResponse`